File: 92-reverse-linked-list-ii/reverse-linked-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head){
        if(head == null) return null;
        ListNode pres = head;
        ListNode prev = null;
        while(pres != null){
            ListNode curr = pres.next;
            pres.next = prev;
            prev = pres;
            pres = curr;
        }
        return prev;
    }
    
    public ListNode reverseBetween(ListNode head, int left, int right) {
        int pos = 1;
        ListNode prev = null;
        ListNode tempOne = null;
        ListNode tempTwo = null;
        ListNode temp = head;
        while(temp != null){
            if(pos == left){
                tempOne = prev;
                tempTwo = temp;
            }
            else if(pos == right){
                ListNode curr = temp.next;
                temp.next = null;
                ListNode node = reverse(tempTwo);
                if(tempOne == null){
                    head = node;
                }
                else tempOne.next = node;
                tempTwo.next = curr;
                break;
            }
            pos++;
            prev = temp;
            temp = temp.next;
        }
        return head;
    }
}

File: 92-reverse-linked-list-ii/reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverse(ListNode* root){
        ListNode* prev = NULL;
        ListNode* curr = root;
        while(curr != NULL){
            ListNode* nextt = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nextt;
        }
        return prev;
    }
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if(left == right) return head;

        ListNode* start = NULL;
        ListNode* end = NULL;
        ListNode* prev = NULL;
        ListNode* temp = head;

        int pos = 1;
        while(temp){
            if(pos == left-1 ){
                prev = temp;
            }
            else if(pos == left){
                start = temp;
            }
            else if(pos == right){
                end = temp->next;
                temp->next = NULL;
                ListNode* reversed = reverse(start);
                if(prev){
                    prev->next = reversed;
                }
                else head = reversed;
                start->next = end;
                break;
            }
            pos++;
            temp = temp->next;
        }
        return head;
    }
};